lte/gateway/python/magma/pipelined/ng_manager/node_state_manager.py

```diff
@@ -0,0 +1,150 @@
+"""
+Copyright 2020 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import logging
+import netifaces
+from typing import NamedTuple, Dict
+from datetime import datetime
+
+from lte.protos.session_manager_pb2 import (
+                          UPFNodeState,
+                          UPFAssociationState,
+                          UPFFeatureSet,
+                          UserPlaneIPResourceSchema
+                        )
+from google.protobuf.timestamp_pb2 import Timestamp
+from magma.pipelined.set_interface_client import set_interface_node_state_association
+from ryu.lib import hub
+
+class NodeStateManager:
+    """
+    This controller manages node state information
+    and reports to SMF.
+    """
+
+    TEID_RANGE_INDICATION = 1
+    TEID_RANGE_VALUE = 0
+    ASSOC_MAX_RETRIES = 40
+
+    GetLocalConfig = NamedTuple(
+        'GetLocalConfig',
+        [('downlink_ip', str), ('node_identifier', str)])
+
+
+    """
+    This controller manages session state information
+    and reports to SMF.
+    """
+    def __init__(self, loop, config):
+        self._loop = loop
+        self.config = self._get_config(config)
+
+        #Fill the node ID for UPF
+        self._node_id=self.config.node_identifier
+
+        self.assoc_message_version = 1
+        self.assoc_message_count = 0
+        self._assoc_established = False
+
+        self.assoc_state = UPFAssociationState.STARTED
+        self._recovery_timestamp = Timestamp()
+        self._recovery_timestamp.FromSeconds(\
```

Comment:
Isn't there something like .GetCurrentTime()

Reply:
Idea was that if UPF and SMF are not on same box then "00:00:00 on 1 January 1900" can be taken as reference time to make them in sync (as per  TS 129 244 V15.8.0, section 8.2.65 Recovery Time Stamp). Same seems to be case with other packet generators for node association.

But Comment is correct in reference that both the process (SessionD and Pipelined) are on same box. Code changes are done as per the comment.